waspc/waspls/test/Wasp/LSP/CompletionTest.hs

```diff
@@ -0,0 +1,170 @@
+            _latestDiagnostics = [],
+            _cst = Just parsedCST
+          }
+      (completionItems, _log) = runLog $ evalStateT (getCompletionsAtPosition cursorPosition) serverState
+      fmtedCompletionItems = map fmtCompletionItem completionItems
+
+      fmtCompletionItem :: LSP.CompletionItem -> String
+      fmtCompletionItem item =
+        unwords
+          [ printf "label={%s}" (show $ item ^. LSP.label),
+            printf "kind={%s}" (show $ item ^. LSP.kind),
+            printf "detail={%s}" (show $ item ^. LSP.detail)
+          ]
+   in "Completion items:\n" ++ unlines (map ("  " <>) fmtedCompletionItems)
+
+-- | Parses a completion test case into a pair of the wasp source code to
+-- run completion on and the position to get the completion list at.
+--
+-- See 'CompletionTestInput' for the format of the test input.
+parseCompletionInput :: CompletionTestInput -> (String, LSP.Position)
+parseCompletionInput (CompletionTestInput waspSourceWithCursor) =
+  if not (hasCompletionTestPreamble waspSourceWithCursor)
+    then error missingCompletionTestPreambleMsg
+    else (waspSourceCode, cursorPosition)
+  where
+    waspSourceCode =
+      unlines . concat $
+        [ linesBeforeCursor,
+          [cursorLineWithCursorRemoved],
+          linesAfterCursor
+        ]
+    cursorPosition = LSP.Position (fromIntegral cursorLineIdx) (fromIntegral cursorColumnIdx)
+
+    linesBeforeCursor = take cursorLineIdx (lines waspSourceWithCursor)
+    -- NOTE: drop cursorLineIdx + 2 so that we skip the "   ^" line after the cursor.
+    linesAfterCursor = drop (cursorLineIdx + 2) (lines waspSourceWithCursor)
+
+    cursorLineWithCursorRemoved =
+      let cursorLine = lines waspSourceWithCursor !! cursorLineIdx
+       in take cursorColumnIdx cursorLine <> drop (cursorColumnIdx + 1) cursorLine
+
+    (cursorLineIdx, cursorColumnIdx) =
+      case listToMaybe $ mapMaybe cursorColumnIdxFromLinePair enumeratedLinePairs of
+        Nothing -> error "parseCompletionInput: found no cursor!"
+        Just v -> v
+    enumeratedLinePairs =
+      zip [0 ..] $ zip (lines waspSourceWithCursor) (drop 1 . lines $ waspSourceWithCursor)
+
+    -- Given two consecutive lines from the test input, along with the line number
+    -- of the first line, look for a cursor in the line pair and find the index
+    -- of the column the cursor is at.
+    cursorColumnIdxFromLinePair :: (Int, (String, String)) -> Maybe (Int, Int)
+    cursorColumnIdxFromLinePair (lineIdx, (line1, line2)) = do
+      cursorColIdx <- elemIndex '|' line1
+      -- Check that the line ends after the column with the | from the previous line
+      guard $ length line2 == cursorColIdx + 1
+      -- Check that ^ is at the same column as | from the previous line
+      guard $ (line2 !! cursorColIdx) == '^'
+      -- Check that only spaces precede the ^
+      guard $ all isSpace $ take cursorColIdx line2
+      return (lineIdx, cursorColIdx)
```

Comment:
Well while we are at it we can go all the way then :D:

```suggestion
    -- Given two consecutive lines from the test input, along with the line index
    -- of the first line, look for a cursor in the line pair and return the index
    -- of the line and column the cursor is at.
    cursorLineIdxAndColumnIdxInLinePair :: (Int, (String, String)) -> Maybe (Int, Int)
    cursorLineIdxAndColumnIdxInLinePair (line1Idx, (line1, line2)) = do
      cursorColIdx <- elemIndex '|' line1
      guard $ line2 == replicate cursorColIdx ' ' <> '^'
      return (lineIdx, cursorColIdx)
```

Ok this is a bit more sensitive because it expects ' ' and not any space but in practice I don't think it matters.

Notice I updated the name of the function and also comments to reflect it also returns line, not just column.

Btw nice idea with `guard`, I forgot about it / never knew it well enough, but is a great fit for this.